### ISSUE TITLE
Creating representations using IBM Watsonx LLMs

### ISSUE THREAD

**Issue by andreswagner** [MaartenGr/BERTopic#2001]
Hi Maarten,

What do you think if we enable BERTopic to create representation by also using Watsonx hosted LLMs like Llama-3-70b, Mixtral-8x7b, Granite and many others to come?

Let me know your thoughts about it to help with the coding

Best,
Andres

**Comment by MaartenGr** [comment]
Although that is currently not implemented, it might be worthwhile to checkout https://github.com/MaartenGr/BERTopic/pull/1451 since that references a broad API-based package that accesses many LLMs. Is that the type of solution that would work for you? If so, I can check whether that might be the best solution here.

**Comment by andreswagner** [comment]
I’m familiar with that package as well. While having a standardized way to use LLMs is beneficial, it does have its downsides, such as the potential for dependency issues. It's not just about accessing various models and vendors; it also involves ensuring that the prompts are optimized for each LLM, requiring small adjustments to the default prompts that create the representations. For the end user, the real value lies in using different LLMs without needing to download any additional packages or tweak any prompts to achieve quality results. Therefore, I propose incorporating Watsonx.ai calls through a standard request library to avoid adding dependency conflicts and adjusting the default prompts for each supported LLM to ensure they deliver good performance right out of the box.

**Comment by andreswagner** [comment]
Hi Maarten,

Well, the beauty of open sourcing your work is that you can have many contributing to keep improving BERTopic. I have been using it since 2021 and I'm happy to become a contributor and getting more people to use it. Regarding LLMs, you're right, is very active right now and every week we see something new gets released. On the other hand, only a few LLMs are available as a service and probably if you try the 80/20 rule, you can count the main Platform Providers with just the fingers of one your hand (check image). Furthermore, most usage comes from a few models that you cant count with the fingers of your other hand. So maybe a interesting strategy might be the following:

1. Blockbuster: To have a few connectors to the Top 5 Platforms and a few tailored made prompts to the Top 5 Models, with the requirements of having this connectors to only use HTTP request so it doesn't add more dependencies conflicts. 
2. Long tail: In case more connectivity is required, then something like LiteLLM + some custom prompt might solve the issue, with the risk of getting dependencies conflict that might need some special care.

On the other hand, if maintaining the Watsonx connector is something that's bugging you, I'm more than happy to keep helping on keeping it updated at BERTopic: to me is much more expensive to keep a fork version updated and sharing that with the rest of the team.

![WhatsApp Image 2024-05-09 at 10 26 25](https://github.com/MaartenGr/BERTopic/assets/13288723/be6a6656-035c-45d7-ad8e-ef6be21109df)

**Comment by andreswagner** [comment]
Hi Maarten,

So what do you think if we start by including the module to have Watsonx called without the need of adding any additional libraries to BERTopic. I have the code already and I've tested it many times with different projects... and is just a few bunch of lines. Are you ok if I do a Pull request with that?

Best,
Andres

**Comment by MaartenGr** [comment]
Yes! A PR would definitely be helpful here, thank you. There's a Ruff PR open that might need to be integrated at some point but since you are essentially making a new `.py` file, you are unlikely to touch existing code much so I don't expect it to be an issue.
